### PR TITLE
feat: Add endpoints for sequence metadata

### LINKED ISSUE #17: Ability to pull cross-references from OEIS
The backend should be able to pull the %Y line from the internal format of a sequence on the OEIS to get the cross-reference, adjacent sequences, and sequence in context data. Might be non-trivial.

### PR DESCRIPTION
**Pull request by gwhitney**
Specifically, api/get_oeis_name_and_values/<OEIS_ID> is guaranteed
  to give the correct official OEIS name of the sequence, and
  api/get_oeis_metadata/<OEIS_ID> which returns the official name,
  cross references, and beack references to the sequence.

  Documents these in the README (as there isn't really any other documentation
  place at the moment).

  Also removes the obsolete TODO.md (it only had one item, which is done, and
  we mostly rely on GitHub issues for todos now anyway).

  Resolves #17.

Basically, this is just a version of #26 compatible with the new resolve to only store data in the database, and that PR should be closed once this is merged.

### CODE REVIEW

**Comment by katestange** [comment]
I got this weird error at the same URL which was previously working (I loaded this url, then tried the "and_name" endpoint in another tab, then came back and reloaded this tab).
Screenshot from 2022-05-02 17-00-37

File: README.md [hunk 1]
```diff
@@ -143,6 +143,71 @@ $ python3 manage.py db upgrade
...
 
+## Endpoints
+
+This section documents all of the endpoints provided by the backscope server.
+All of them return JSON data with the specified keys and values. Also, every
+endpoint includes the key 'id' with value the OEIS id for the sake of verifying
+that it is the data as requested. In case of an OEIS_ID that does not match
+anything in the OEIS, an error string is returned. Note that the angle brackets
+<> in the URLS indicate where subsitutions are made, they should not be present
+in the URLs actually used.
```
Comment (katestange) [inline README.md:154]: comma splice :)

File: README.md [hunk 2]
```diff
@@ -143,6 +143,71 @@ $ python3 manage.py db upgrade
...
+
+Also note that if any of the requests are made for a given sequence, then the
+back end will in the background obtain all of the data necessary to respond
+to all of the endpoints for future requests concerning that sequence without
+going back to the OEIS.
```
Comment (katestange) [inline README.md:159]: If you make the same time-intensive request in quick succession, so the second request comes in before the backend is done pulling all the info and putting it in the database, is the backend smart enough not to start a long series of requests redundantly twice? Will this cause trouble?

**Comment by katestange** [comment]
Another comment:  could we put a full example URL for a safe (not too many crossrefs) sequence in the readme, so people can click something and see it in action?  I know the base URL may change, but that would work, for example, on my system.  Or is this bad practice?

**Comment by gwhitney** [comment]
> I got this weird error at the same URL

Not too much I can do about this directly with the limited info. I will put in a trap for a non-integer where that happened and issue what will hopefully be a more informative error in case it happens again.

File: README.md [hunk 3]
```diff
@@ -143,6 +143,71 @@ $ python3 manage.py db upgrade
...
+if already known to backscope, or a temporary name if not.
+
+#### Key: values
+
+An array of _strings_ (of digits) giving the first COUNT values of the sequence
+with id OEIS_ID. Since some sequence values correspond to extremely large
+numbers, strings are used to avoid the limitations of any particular numeric
+datatype.
+
+### URL: api/get_oeis_name_and_values/<OEIS_ID>
```
[the remark this answers is not in the document]
Comment (gwhitney) [reply README.md:178]: No and no. In this second endpoint, I figured if you want to take the time for the name, we may as well send you all the terms because in practice there are rarely more than 10000 which just isn't a problem to send. And the third endpoint does not send any values, so there was no need for a COUNT. If you want me to change the second endpoint so it does take a count, let me know.

File: README.md (hunk 1 above)
Comment (gwhitney) [reply README.md:154]: Excellent catch, will fix.

File: README.md (hunk 2 above)
Comment (gwhitney) [reply README.md:159]: Nope,the backend is currently not smart enough. All those redundant requests will occur. The database should be smart enough, though, for things to end up in an OK state, but all of the work will be done twice. I am not really sure how to fix this, and I don't particularly relish figuring out how. I guess I could use something in the database as a sentinel that fetching is going on. How much are you worried about the potentially duplicated work? I.e., should we just file an issue that there is the possibility of such duplicated work, or do you want me to try to forestall it before this is merged.

**Comment by gwhitney** [comment]
> Another comment: could we put a full example URL for a safe (not too many crossrefs) sequence in the readme, so people can click something and see it in action? I know the base URL may change, but that would work, for example, on my system. Or is this bad practice?

Sure, I will add an example to each of the URL sections with a not-too-big sequence ID.

**Comment by gwhitney** [comment]
OK, I've made the fixes described above. Feel free to request further action on the items you brought up that I responded to merely with an explanation, so far.

File: README.md (hunk 2 above)
Comment (katestange) [reply README.md:160]: I mainly wanted to make sure I understand the state of affairs.  I don't feel a need to do anything about it in this PR, or anytime soon, since I don't see it being a major issue.  But do you have a suggestion on the most appropriate place to document that this is the case?  For example, is there a spot in the code we can clarify that this may occur, for some future reader of the codebase?  That may be more useful than opening an issue, if our intention with the issue is mainly to ignore the issue.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (katestange) [reply README.md:182]: Ok, so it's clear the third endpoint doesn't need it, true.  So, just to clarify, you're saying for the first endpoint, the only reason you can select a number of terms is so you can be ultrafast about it, and the second endpoint isn't ultrafast in any case, so why bother?  I suppose that makes sense.

[the remark this answers is not in the document]
Comment (gwhitney) [reply README.md:182]: Well, that was exactly my rationale for not including the count in endpoint 2. If you think it's better to have the count in endpoint 2 for consistency, just let me know.

[the remark this answers is not in the document]
Comment (katestange) [reply README.md:182]: Hmm, do you think someone programming on the front end might find it annoying that there's a difference?  In your experience as a programmer?  I mean, if you're building a URL and you need to make a case statement...

File: README.md (hunk 2 above)
Comment (gwhitney) [reply README.md:160]: I will add a comment in the appropriate place in the code, but I recommend we open an issue as well. It can always be closed if we want to "clean up" the issues at some point, but even if so, it will document that we considered the point and it can be searched for in the future. If you have no objection, let me know and I will file one.

File: README.md [hunk 4]
```diff
@@ -143,6 +144,81 @@ $ python3 manage.py db upgrade
...
+to all of the endpoints for future requests concerning that sequence without
+going back to the OEIS.
+
+### URL: api/get_oeis_values/<OEIS_ID>/<COUNT>
+
+This is the most rapid endpoint, it makes at most one request to the OEIS server
+(and only if the OEIS_ID has not previously been requested). If you are running
+the server to test it on your local host, a full URL would be
+`http://127.0.0.1:5000/api/get_oeis_values/A000030/50` which will return the
+first digits of the numbers 0 through 49.
```
Comment (katestange) [inline README.md:168]: Maybe add something along the lines of:  By limiting the COUNT to a small value, you can further speed up this request, while initiating the full pull of data in the background, so that the full data will be available immediately at a later request.  Kind of a hint to programmers of visualizers.  (If that's true, which is what I understood from our other conversation.)

File: README.md [hunk 5]
```diff
@@ -143,6 +144,81 @@ $ python3 manage.py db upgrade
...
+
+An array of _strings_ (of digits) giving the first COUNT values of the sequence
+with id OEIS_ID. Since some sequence values correspond to extremely large
+numbers, strings are used to avoid the limitations of any particular numeric
+datatype.
+
+### URL: api/get_oeis_name_and_values/<OEIS_ID>
+
+Potentially a bit slower than the above URL, it may make an extra request to
+ensure that the name is correct. If you are running the server on your local
```
Comment (katestange) [inline README.md:185]: I'm not sure if this is strictly a comma splice, but something is awkward about it.

File: README.md [hunk 6]
```diff
@@ -143,6 +144,81 @@ $ python3 manage.py db upgrade
...
+host, a full URL would be `http://127.0.0.1:5000/api/get_oeis_values/A003173`
+which will return the nine Heegner numbers and their full name as an OEIS
+sequence (basically, the name describes what a Heegner number is).
```
Comment (katestange) [inline README.md:188]: I love that you picked specific sequences and mentioned what they are. :)

File: README.md [hunk 7]
```diff
@@ -143,6 +144,81 @@ $ python3 manage.py db upgrade
...
+
+#### Key: name
+
+A string giving the official name of the OEIS sequence with id OEIS_ID
```
Comment (katestange) [inline README.md:192]: missing period.

File: README.md [hunk 8]
```diff
@@ -143,6 +144,81 @@ $ python3 manage.py db upgrade
...
+may make hundreds of requests to the OEIS to generate all of the back
+references to the sequence. If you are running the server on your local
+machine, a full URL would be
+`http://127.0.0.1:5000/api/get_oeis_metadata/A028444` which will show the full
+name of the Busy Beaver sequence, the one text line of sequences it
+references, and the IDs of the ten sequences that refer to it.
+
+#### Key: name
+
+A string giving the official name of the OEIS sequence with id OEIS_ID
```
Comment (katestange) [inline README.md:211]: missing period.

File: README.md (hunk 6 above)
Comment (katestange) [reply README.md:188]: Also, the URL is wrong. It should say "get_oeis_name_and_values"

**Comment by katestange** [comment]
> > I got this weird error at the same URL
> 
> Not too much I can do about this directly with the limited info. I will put in a trap for a non-integer where that happened and issue what will hopefully be a more informative error in case it happens again.

I was not able to reproduce it this time around, although I tried to do a bit of stress-testing and managed the attached screenshot, which means things are working as described. :)

Screenshot from 2022-05-02 22-13-10

**Comment by katestange** [comment]
Ok, here's a mystery for you.  When I look at the markup directly, I don't see "COUNT" in the first endpoint.  But when I look in the diff of changes on github, or in my local directory, I see it.  Am I losing my mind?

Screenshot from 2022-05-02 22-16-29

Screenshot from 2022-05-02 22-17-00

File: README.md (hunk 2 above)
Comment (katestange) [reply README.md:160]: Ok, that sounds good to me.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (gwhitney) [reply README.md:182]: Hard to know. Right now, there's just one way the front end obtains sequences from the back end, it's the OEISSequenceTemplate. If you're writing a visualizer you pretty much don't directly access sequence data, you just use a Sequence object. So I don't think many visualizer writers are even going to worry about how they are getting their sequence data.

File: README.md [hunk 9]
```diff
@@ -143,6 +144,81 @@ $ python3 manage.py db upgrade
...
+endpoint includes the key 'id' with value the OEIS id for the sake of verifying
+that it is the data as requested. In case of an OEIS_ID that does not match
+anything in the OEIS, an error string is returned. Note that the angle brackets
+<> in the URLS indicate where subsitutions are made; they should not be present
+in the URLs actually used.
+
+Also note that if any of the requests are made for a given sequence, then the
+back end will in the background obtain all of the data necessary to respond
+to all of the endpoints for future requests concerning that sequence without
+going back to the OEIS.
```
[the remark this answers is not in the document]
Comment (gwhitney) [reply README.md:160]: Yup will do.

File: README.md (hunk 4 above)
Comment (gwhitney) [reply README.md:168]: Well, the count is there in case it could speed things up, but in the current implementation, it doesn't actually. The thing is that everything is so fast even when you get all 10000 entries that it doesn't much matter. I just left the count because it was already there, and in principle maybe it could be faster, and if we ever need to, we could try to implement something faster. As I mentioned above, visualizers mostly just use a Sequence object, and we have just one kind of sequence object that accesses the backend, so this is not something that most visualizer writers will even have to worry about. So I am inclined to leave this bit as is, but let me know if you still want more commentary.

File: README.md (hunk 5 above)
Comment (gwhitney) [reply README.md:185]: Love your attention to exposition. Will fix.

File: README.md (hunk 7 above)
Comment (gwhitney) [reply README.md:192]: Roger

File: README.md (hunk 8 above)
Comment (gwhitney) [reply README.md:211]: Ditto

**Comment by gwhitney** [comment]
> I was not able to reproduce it this time around

Ah but you did. That must be the message that was causing the error before. So my trap is working, and revealing the issue. So yes, it seems as though the OEIS is detecting that we generate a lot of hits and protecting itself, and so if one just waits and tries again it works. That state of affairs seems OK to me; I think we can just go ahead with this scheme and if any difficulties seem to arise in what actually ends up in the database we can track them down when it happens. Sound OK?

**Comment by gwhitney** [comment]
> Ok, here's a mystery for you.

Must be something about how github markdown works. I will try to fix in the next commit.

**Comment by gwhitney** [comment]
Thanks for the excellent review. I think I have taken care of everything as discussed, although there was one point where I proposed leaving the documentation as is. So let me know if that's OK, and if everything else seems resolved.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (katestange) [reply README.md]: Ah, I see, right, of course that's the right way to handle it.  Ok, I agree, we needn't add a COUNT anywhere else.

File: README.md (hunk 4 above)
Comment (katestange) [reply README.md:170]: Ok, let's leave it.

**Comment by katestange** [comment]
> > I was not able to reproduce it this time around
> 
> Ah but you did. That must be the message that was causing the error before. So my trap is working, and revealing the issue. So yes, it seems as though the OEIS is detecting that we generate a lot of hits and protecting itself, and so if one just waits and tries again it works. That state of affairs seems OK to me; I think we can just go ahead with this scheme and if any difficulties seem to arise in what actually ends up in the database we can track them down when it happens. Sound OK?

Oh cool!  Ok, great!  Yes, the error message is plenty informative now.  One more question, though:  would it be good to trap this and prevent further requests for a period of time when this error arises (in this PR or as a separate issue for later)?

**Comment by gwhitney** [comment]
> One more question, though: would it be good to trap this and prevent further requests for a period of time when this error arises (in this PR or as a separate issue for later)?

Yes, that is a very good idea. I think it is outside the scope of this PR, but absolutely we should file an issue for this. I don't offhand know how to implement something like that (I'm not really experienced with concurrent/multithreaded programming) but it's well worth looking into, particularly from the "don't upset the folks you eventually want to work with" point of view.

**Comment by katestange** [comment]
> > One more question, though: would it be good to trap this and prevent further requests for a period of time when this error arises (in this PR or as a separate issue for later)?
> 
> Yes, that is a very good idea. I think it is outside the scope of this PR, but absolutely we should file an issue for this. I don't offhand know how to implement something like that (I'm not really experienced with concurrent/multithreaded programming) but it's well worth looking into, particularly from the "don't upset the folks you eventually want to work with" point of view.

Ok, great, then let's file that issue (and the other one mentioned in this conversation).  The only other outstanding thing then for me on this is the missing <COUNT> in the markup... let's face it, anyone taking an interest will read it via github, and that's very confusing.  Is there a workaround?

**Comment by gwhitney** [comment]
> the missing COUNT in the markup

Did my fix not work? Can you send another screenshot? We should definitely fix this before merging -- the README needs to look right.

**Comment by katestange** [comment]
> > the missing COUNT in the markup
> 
> Did my fix not work? Can you send another screenshot? We should definitely fix this before merging -- the README needs to look right.

Oh it's fixed!  I needed to refresh something.
